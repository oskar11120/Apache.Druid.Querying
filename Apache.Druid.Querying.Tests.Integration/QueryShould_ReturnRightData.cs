using FluentAssertions;
using Snapshooter;
using Snapshooter.NUnit;
using System.Globalization;
using System.Runtime.CompilerServices;
using static Apache.Druid.Querying.Tests.Integration.ServiceProvider;

namespace Apache.Druid.Querying.Tests.Integration;

internal static class TestData
{
    public static readonly DateTimeOffset T0 = DateTimeOffset.Parse("2016-06-27T00:00:00.000Z", CultureInfo.InvariantCulture, DateTimeStyles.AssumeUniversal);
    public static readonly Interval Interval = new(T0, T0.AddDays(1));

    public static TQuery DefaultInterval<TQuery>(this TQuery query)
        where TQuery : IQueryWith.Intervals
        => query
            .Interval(Interval);
}

internal class QueryShould_ReturnRightData
{
    private static async Task VerifyMatch<TSource, TResult, TContext>(
        DataSource<TSource> dataSource,
        IQueryWithSource<TSource>.AndResult<TResult>.AndDeserializationAndTruncatedResultHandling<TContext> query,
        [CallerArgumentExpression(nameof(query))] string snapshotNameSuffix = "")
        where TContext : new()
    {
        var spanshotName = Snapshot.FullName(new SnapshotNameExtension(snapshotNameSuffix));
        var json = dataSource.MapQueryToJson(query).ToString();
        await TestContext.Out.WriteLineAsync(json);
        void Match(List<TResult> results)
            => Snapshot.Match(results, spanshotName, options => options.IgnoreField("[:].SegmentId"));
        var results = await dataSource
            .ExecuteQuery(query)
            .ToListAsync();
        Match(results);

        var withNoTruncatedResultHandling = await dataSource
            .ExecuteQuery(query, onTruncatedResultsQueryRemaining: false)
            .ToListAsync();
        Match(withNoTruncatedResultHandling);
    }

    private static Task VerifyMatch<TResult, TContext>(
        IQueryWithSource<Edit>.AndResult<TResult>.AndDeserializationAndTruncatedResultHandling<TContext> query,
        [CallerArgumentExpression(nameof(query))] string snapshotNameSuffix = "")
        where TContext : new()
        => VerifyMatch(Wikipedia.Edits, query, snapshotNameSuffix);

    [Test]
    public async Task Wrapped()
    {
        var first = new Query<Edit>
            .TimeSeries
            .WithNoVirtualColumns
            .WithAggregations<TimeSeriesAggregations>()
            .Aggregations(type => new(
                type.Count(),
                type.Sum(edit => edit.Added)))
            .DefaultInterval()
            .Granularity(Granularity.FifteenMinutes)
            .Context(new() { SkipEmptyBuckets = true });
        var wrapped = Wikipedia
            .Edits
            .ToQueryDataSource(first);
        var query = new Query<WithTimestamp<TimeSeriesAggregations>>
            .Scan()
            .Interval(TestData.Interval)
            .Filter(type => type.Not(type.Selector(data => data.Value.Count, 6)))
            .Limit(1000);
        await VerifyMatch(wrapped, query, string.Empty);
    }

    public enum JoinTestCase
    {
        LatestEditTimestampsPerCountry,
        LatestEditsPerCountry,
        Both
    }
    private record Country(string Code, string FullName);
    [TestCase(JoinTestCase.LatestEditTimestampsPerCountry)]
    [TestCase(JoinTestCase.LatestEditsPerCountry)]
    [TestCase(JoinTestCase.Both)]
    public async Task Join(JoinTestCase @case)
    {
        var latestEditTimestampsPerCountry_query = new Query<Edit>
            .GroupBy<string>
            .WithNoVirtualColumns
            .WithAggregations<DateTimeOffset>()
            .DefaultInterval()
            .Granularity(Granularity.Day)
            .Dimensions(type => type.Default(data => data.CountryName))
            .Aggregations(type => type.Last(data => data.Timestamp));
        var latestEditTimestampsPerCountry_dataSource = Wikipedia.Edits.ToQueryDataSource(latestEditTimestampsPerCountry_query);
        var lastestEditsPerCountry_dataSource = Wikipedia.Edits.InnerJoin(
            latestEditTimestampsPerCountry_dataSource,
            data => $"{data.Left.CountryName} == {data.Right.Value.Dimensions} && {data.Left.Timestamp} == {data.Right.Value.Aggregations}");
        var edits_query = new Query<InnerJoinData<Edit, WithTimestamp<Dimensions_Aggregations<string, DateTimeOffset>>>>
            .Scan()
            .DefaultInterval();
        await (@case switch
        {
            JoinTestCase.LatestEditTimestampsPerCountry => VerifyMatch(Wikipedia.Edits, latestEditTimestampsPerCountry_query, @case.ToString()),
            JoinTestCase.LatestEditsPerCountry => VerifyMatch(lastestEditsPerCountry_dataSource, edits_query, @case.ToString()),
            JoinTestCase.Both => VerifyBoth(),
            _ => throw new NotSupportedException()
        });

        async Task VerifyBoth()
        {
            var fromEdits = await lastestEditsPerCountry_dataSource
                .ExecuteQuery(edits_query)
                .Select(result => result.Value.Right.Value)
                .OrderBy(result => result.Aggregations)
                .ToArrayAsync();
            var fromTimestamps = await Wikipedia.Edits
                .ExecuteQuery(latestEditTimestampsPerCountry_query)
                .Select(result => result.Value)
                .Where(result => result.Dimensions is not null)
                .OrderBy(result => result.Aggregations)
                .ToArrayAsync();
            fromEdits.Should().BeEquivalentTo(fromTimestamps);
        }
    }

    [Test]
    public async Task Union()
    {
        var union = Wikipedia
            .Edits
            .Union(Wikipedia
                .Edits);
        var query = new Query<Union<Edit, Edit>>
            .Scan()
            .Interval(TestData.Interval)
            .Limit(10);
        await VerifyMatch(union, query, string.Empty);
    }

    private sealed record InlineData(string Word, [property: DataSourceColumn("Integer")] int Number);
    [Test]
    public async Task Inline()
    {
        var query = new Query<InlineData>
            .Scan()
            .Interval(new(default, DateTimeOffset.MaxValue))
            .Filter(type => type.In(
                data => data.Word,
                new[] { "zero" }));
        var inline = Wikipedia
            .Inline(new InlineData[]
            {
                new("zero", 1),
                new("one", 2)
            });
        var result = await inline
            .ExecuteQuery(query)
            .ToListAsync();
        result
            .Single()
            .Value
            .Should()
            .Be(new InlineData("zero", 1));
    }

    [Test]
    public async Task Scan()
    {
        var first3Us = new Query<Edit>
            .Scan()
            .DefaultInterval()
            .Filter(type => type.Selector(
                edit => edit.CountryIsoCode,
                "US"))
            .Limit(3);
        await VerifyMatch(first3Us);
    }

    [Test]
    public async Task GroupBy()
    {
        var countsPerCountry = new Query<Edit>
            .GroupBy<string>
            .WithNoVirtualColumns
            .WithAggregations<int>()
            .DefaultInterval()
            .Dimensions(type => type.Default(edit => edit.CountryName))
            .Aggregations(type => type.Count())
            .Granularity(Granularity.EightHours);

        var countsPerCountryOrderedByCount = countsPerCountry
            .LimitSpec(limit: 100, offset: 1, columns => columns.OrderBy(data => data.Aggregations, OrderDirection.Descending, SortingOrder.Numeric));
        await VerifyMatch(countsPerCountryOrderedByCount);

        var countsPerValidCountry = countsPerCountry
            .Filter(type => type.Not(type.Selector(edit => edit.CountryName, string.Empty)));
        await VerifyMatch(countsPerValidCountry);
    }

    [Test]
    public async Task TopN()
    {
        var countsPerPage = new Query<Edit>
            .TopN<string>
            .WithNoVirtualColumns
            .WithAggregations<int>()
            .Dimension(type => type.Default(edit => edit.Page))
            .Threshold(10)
            .Aggregations(type => type.Count())
            .Metric(type => type.Numeric(data => data.Aggregations))
            .DefaultInterval()
            .Granularity(Granularity.Hour);
        await VerifyMatch(countsPerPage);

        var usAnonymousCountsPerPage = countsPerPage
            .Filter(type => type.And(
                type.Selector(edit => edit.IsAnonymous, true),
                type.Selector(edit => edit.CountryIsoCode, "US")));
        await VerifyMatch(usAnonymousCountsPerPage);
    }

    private record TimeSeriesAggregations(int Count, int TotalAdded);
    private record TimeSeriesPostAggrgations(double AverageAdded);
    [Test]
    public async Task TimeSeries()
    {
        var lineStatisticsPerHour = new Query<Edit>
            .TimeSeries
            .WithNoVirtualColumns
            .WithAggregations<TimeSeriesAggregations>
            .WithPostAggregations<TimeSeriesPostAggrgations>()
            .Order(OrderDirection.Descending)
            .Aggregations(type => new(
                type.Count(),
                type.Sum(edit => edit.Added)))
            .PostAggregations(type => new(type.Arithmetic(
                ArithmeticFunction.Divide,
                type.FieldAccess(aggregations => aggregations.TotalAdded),
                type.FieldAccess(aggregations => aggregations.Count))))
            .Filter(type => type.Selector(edit => edit.CountryIsoCode, "US"))
            .DefaultInterval()
            .Granularity(Granularity.Hour)
            .Context(new QueryContext.TimeSeries() { SkipEmptyBuckets = true });
        await VerifyMatch(lineStatisticsPerHour);

        var inEuWarsawTime = lineStatisticsPerHour
            .Granularity(Granularity.Hour, "Europe/Warsaw");
        await VerifyMatch(inEuWarsawTime);
    }


    private sealed record GroupByBooleans(bool Robot, bool New);
    [TestCase(0)]
    [TestCase(1)]
    public async Task WithDataPropertiesAccessedByInterface(int i)
    {
        async Task Test<TEdit>(DataSource<TEdit> dataSource) where TEdit : IEditBooleans
        {
            var one = new Query<TEdit>
                .Scan()
                .DefaultInterval()
                .Filter(type => type.And(
                    type.Selector(data => data.IsNew, true),
                    type.Selector(data => data.Robot, true)))
                .Limit(100);
            if (i is 0)
                await VerifyMatch(dataSource, one);

            var two = new Query<TEdit>
                .GroupBy<GroupByBooleans>
                .WithNoVirtualColumns
                .WithAggregations<int>()
                .DefaultInterval()
                .Granularity(Granularity.Hour)
                .Dimensions(type => new(
                    type.Default(data => data.Robot),
                    type.Default(data => data.IsNew)))
                .Aggregations(type => type.Count());
            if (i is 1)
                await VerifyMatch(dataSource, two);
        }

        await Test(Wikipedia.Edits);
    }

    interface IIotMeasurement
    {
        Guid ObjectId { get; }
        double Value { get; }
    }
}
