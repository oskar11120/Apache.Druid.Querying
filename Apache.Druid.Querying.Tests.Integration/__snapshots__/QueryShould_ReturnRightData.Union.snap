[
  {
    "SegmentId": "wikipedia_2016-06-27T00:00:00.000Z_2016-06-28T00:00:00.000Z_2024-02-17T12:37:41.909Z",
    "Value": {
      "First": {
        "Timestamp": "2016-06-27T00:00:11.08+00:00",
        "IsRobot": true,
        "Channel": "#sv.wikipedia",
        "Flags": "NB",
        "IsUnpatrolled": false,
        "Page": "Salo Toraut",
        "DiffUri": "https://sv.wikipedia.org/w/index.php?oldid=36099284&rcid=89369918",
        "Added": 31,
        "Comment": "Botskapande Indonesien omdirigering",
        "CommentLength": 35,
        "IsNew": true,
        "IsMinor": false,
        "Delta": 31,
        "IsAnonymous": false,
        "User": "Lsjbot",
        "DeltaBucket": 0,
        "Deleted": 0,
        "Namespace": "Main",
        "CityName": null,
        "CountryName": null,
        "RegionIsoCode": null,
        "MetroCode": null,
        "CountryIsoCode": null,
        "RegionName": null
      },
      "Second": {
        "Timestamp": "2016-06-27T00:00:11.08+00:00",
        "IsRobot": true,
        "Channel": "#sv.wikipedia",
        "Flags": "NB",
        "IsUnpatrolled": false,
        "Page": "Salo Toraut",
        "DiffUri": "https://sv.wikipedia.org/w/index.php?oldid=36099284&rcid=89369918",
        "Added": 31,
        "Comment": "Botskapande Indonesien omdirigering",
        "CommentLength": 35,
        "IsNew": true,
        "IsMinor": false,
        "Delta": 31,
        "IsAnonymous": false,
        "User": "Lsjbot",
        "DeltaBucket": 0,
        "Deleted": 0,
        "Namespace": "Main",
        "CityName": null,
        "CountryName": null,
        "RegionIsoCode": null,
        "MetroCode": null,
        "CountryIsoCode": null,
        "RegionName": null
      }
    }
  },
  {
    "SegmentId": "wikipedia_2016-06-27T00:00:00.000Z_2016-06-28T00:00:00.000Z_2024-02-17T12:37:41.909Z",
    "Value": {
      "First": {
        "Timestamp": "2016-06-27T00:00:17.457+00:00",
        "IsRobot": false,
        "Channel": "#ja.wikipedia",
        "Flags": "",
        "IsUnpatrolled": false,
        "Page": "利用者:ワーナー成増/放送ウーマン賞",
        "DiffUri": "https://ja.wikipedia.org/w/index.php?diff=60239890&oldid=60239620",
        "Added": 125,
        "Comment": "70年代",
        "CommentLength": 4,
        "IsNew": false,
        "IsMinor": false,
        "Delta": 125,
        "IsAnonymous": false,
        "User": "ワーナー成増",
        "DeltaBucket": 100,
        "Deleted": 0,
        "Namespace": "利用者",
        "CityName": null,
        "CountryName": null,
        "RegionIsoCode": null,
        "MetroCode": null,
        "CountryIsoCode": null,
        "RegionName": null
      },
      "Second": {
        "Timestamp": "2016-06-27T00:00:17.457+00:00",
        "IsRobot": false,
        "Channel": "#ja.wikipedia",
        "Flags": "",
        "IsUnpatrolled": false,
        "Page": "利用者:ワーナー成増/放送ウーマン賞",
        "DiffUri": "https://ja.wikipedia.org/w/index.php?diff=60239890&oldid=60239620",
        "Added": 125,
        "Comment": "70年代",
        "CommentLength": 4,
        "IsNew": false,
        "IsMinor": false,
        "Delta": 125,
        "IsAnonymous": false,
        "User": "ワーナー成増",
        "DeltaBucket": 100,
        "Deleted": 0,
        "Namespace": "利用者",
        "CityName": null,
        "CountryName": null,
        "RegionIsoCode": null,
        "MetroCode": null,
        "CountryIsoCode": null,
        "RegionName": null
      }
    }
  },
  {
    "SegmentId": "wikipedia_2016-06-27T00:00:00.000Z_2016-06-28T00:00:00.000Z_2024-02-17T12:37:41.909Z",
    "Value": {
      "First": {
        "Timestamp": "2016-06-27T00:00:34.959+00:00",
        "IsRobot": false,
        "Channel": "#en.wikipedia",
        "Flags": "",
        "IsUnpatrolled": false,
        "Page": "Bailando 2015",
        "DiffUri": "https://en.wikipedia.org/w/index.php?diff=727144213&oldid=727144184",
        "Added": 2,
        "Comment": "/* Scores */",
        "CommentLength": 12,
        "IsNew": false,
        "IsMinor": false,
        "Delta": 2,
        "IsAnonymous": true,
        "User": "181.230.118.178",
        "DeltaBucket": 0,
        "Deleted": 0,
        "Namespace": "Main",
        "CityName": "Buenos Aires",
        "CountryName": "Argentina",
        "RegionIsoCode": "C",
        "MetroCode": null,
        "CountryIsoCode": "AR",
        "RegionName": "Buenos Aires F.D."
      },
      "Second": {
        "Timestamp": "2016-06-27T00:00:34.959+00:00",
        "IsRobot": false,
        "Channel": "#en.wikipedia",
        "Flags": "",
        "IsUnpatrolled": false,
        "Page": "Bailando 2015",
        "DiffUri": "https://en.wikipedia.org/w/index.php?diff=727144213&oldid=727144184",
        "Added": 2,
        "Comment": "/* Scores */",
        "CommentLength": 12,
        "IsNew": false,
        "IsMinor": false,
        "Delta": 2,
        "IsAnonymous": true,
        "User": "181.230.118.178",
        "DeltaBucket": 0,
        "Deleted": 0,
        "Namespace": "Main",
        "CityName": "Buenos Aires",
        "CountryName": "Argentina",
        "RegionIsoCode": "C",
        "MetroCode": null,
        "CountryIsoCode": "AR",
        "RegionName": "Buenos Aires F.D."
      }
    }
  },
  {
    "SegmentId": "wikipedia_2016-06-27T00:00:00.000Z_2016-06-28T00:00:00.000Z_2024-02-17T12:37:41.909Z",
    "Value": {
      "First": {
        "Timestamp": "2016-06-27T00:00:36.027+00:00",
        "IsRobot": false,
        "Channel": "#en.wikipedia",
        "Flags": "M",
        "IsUnpatrolled": false,
        "Page": "Richie Rich's Christmas Wish",
        "DiffUri": "https://en.wikipedia.org/w/index.php?diff=727144214&oldid=716794477",
        "Added": 0,
        "Comment": "standard term is [[title character]]",
        "CommentLength": 36,
        "IsNew": false,
        "IsMinor": true,
        "Delta": -2,
        "IsAnonymous": false,
        "User": "JasonAQuest",
        "DeltaBucket": -100,
        "Deleted": 2,
        "Namespace": "Main",
        "CityName": null,
        "CountryName": null,
        "RegionIsoCode": null,
        "MetroCode": null,
        "CountryIsoCode": null,
        "RegionName": null
      },
      "Second": {
        "Timestamp": "2016-06-27T00:00:36.027+00:00",
        "IsRobot": false,
        "Channel": "#en.wikipedia",
        "Flags": "M",
        "IsUnpatrolled": false,
        "Page": "Richie Rich's Christmas Wish",
        "DiffUri": "https://en.wikipedia.org/w/index.php?diff=727144214&oldid=716794477",
        "Added": 0,
        "Comment": "standard term is [[title character]]",
        "CommentLength": 36,
        "IsNew": false,
        "IsMinor": true,
        "Delta": -2,
        "IsAnonymous": false,
        "User": "JasonAQuest",
        "DeltaBucket": -100,
        "Deleted": 2,
        "Namespace": "Main",
        "CityName": null,
        "CountryName": null,
        "RegionIsoCode": null,
        "MetroCode": null,
        "CountryIsoCode": null,
        "RegionName": null
      }
    }
  },
  {
    "SegmentId": "wikipedia_2016-06-27T00:00:00.000Z_2016-06-28T00:00:00.000Z_2024-02-17T12:37:41.909Z",
    "Value": {
      "First": {
        "Timestamp": "2016-06-27T00:00:46.874+00:00",
        "IsRobot": true,
        "Channel": "#sh.wikipedia",
        "Flags": "MB",
        "IsUnpatrolled": false,
        "Page": "El Olivo, Ascensión",
        "DiffUri": "https://sh.wikipedia.org/w/index.php?diff=29388551&oldid=29282558&rcid=34712908",
        "Added": 0,
        "Comment": "Bot: Automatska zamjena teksta  (-[[Administrativna podjela Meksika|Admin]] +[[Administrativna podjela Meksika|Admi]])",
        "CommentLength": 118,
        "IsNew": false,
        "IsMinor": true,
        "Delta": -1,
        "IsAnonymous": false,
        "User": "Kolega2357",
        "DeltaBucket": -100,
        "Deleted": 1,
        "Namespace": "Main",
        "CityName": null,
        "CountryName": null,
        "RegionIsoCode": null,
        "MetroCode": null,
        "CountryIsoCode": null,
        "RegionName": null
      },
      "Second": {
        "Timestamp": "2016-06-27T00:00:46.874+00:00",
        "IsRobot": true,
        "Channel": "#sh.wikipedia",
        "Flags": "MB",
        "IsUnpatrolled": false,
        "Page": "El Olivo, Ascensión",
        "DiffUri": "https://sh.wikipedia.org/w/index.php?diff=29388551&oldid=29282558&rcid=34712908",
        "Added": 0,
        "Comment": "Bot: Automatska zamjena teksta  (-[[Administrativna podjela Meksika|Admin]] +[[Administrativna podjela Meksika|Admi]])",
        "CommentLength": 118,
        "IsNew": false,
        "IsMinor": true,
        "Delta": -1,
        "IsAnonymous": false,
        "User": "Kolega2357",
        "DeltaBucket": -100,
        "Deleted": 1,
        "Namespace": "Main",
        "CityName": null,
        "CountryName": null,
        "RegionIsoCode": null,
        "MetroCode": null,
        "CountryIsoCode": null,
        "RegionName": null
      }
    }
  },
  {
    "SegmentId": "wikipedia_2016-06-27T00:00:00.000Z_2016-06-28T00:00:00.000Z_2024-02-17T12:37:41.909Z",
    "Value": {
      "First": {
        "Timestamp": "2016-06-27T00:00:56.913+00:00",
        "IsRobot": false,
        "Channel": "#en.wikipedia",
        "Flags": "M",
        "IsUnpatrolled": false,
        "Page": "Blowback (intelligence)",
        "DiffUri": "https://en.wikipedia.org/w/index.php?diff=727144242&oldid=726157098",
        "Added": 76,
        "Comment": "linking to other wikipedia pages",
        "CommentLength": 32,
        "IsNew": false,
        "IsMinor": true,
        "Delta": 76,
        "IsAnonymous": false,
        "User": "Brokenshardz",
        "DeltaBucket": 0,
        "Deleted": 0,
        "Namespace": "Main",
        "CityName": null,
        "CountryName": null,
        "RegionIsoCode": null,
        "MetroCode": null,
        "CountryIsoCode": null,
        "RegionName": null
      },
      "Second": {
        "Timestamp": "2016-06-27T00:00:56.913+00:00",
        "IsRobot": false,
        "Channel": "#en.wikipedia",
        "Flags": "M",
        "IsUnpatrolled": false,
        "Page": "Blowback (intelligence)",
        "DiffUri": "https://en.wikipedia.org/w/index.php?diff=727144242&oldid=726157098",
        "Added": 76,
        "Comment": "linking to other wikipedia pages",
        "CommentLength": 32,
        "IsNew": false,
        "IsMinor": true,
        "Delta": 76,
        "IsAnonymous": false,
        "User": "Brokenshardz",
        "DeltaBucket": 0,
        "Deleted": 0,
        "Namespace": "Main",
        "CityName": null,
        "CountryName": null,
        "RegionIsoCode": null,
        "MetroCode": null,
        "CountryIsoCode": null,
        "RegionName": null
      }
    }
  },
  {
    "SegmentId": "wikipedia_2016-06-27T00:00:00.000Z_2016-06-28T00:00:00.000Z_2024-02-17T12:37:41.909Z",
    "Value": {
      "First": {
        "Timestamp": "2016-06-27T00:00:58.599+00:00",
        "IsRobot": true,
        "Channel": "#pl.wikipedia",
        "Flags": "NB",
        "IsUnpatrolled": false,
        "Page": "Kategoria:Dyskusje nad usunięciem artykułu zakończone bez konsensusu − lipiec 2016",
        "DiffUri": "https://pl.wikipedia.org/w/index.php?oldid=46204477&rcid=68522573",
        "Added": 270,
        "Comment": "utworzenie kategorii",
        "CommentLength": 20,
        "IsNew": true,
        "IsMinor": false,
        "Delta": 270,
        "IsAnonymous": false,
        "User": "Beau.bot",
        "DeltaBucket": 200,
        "Deleted": 0,
        "Namespace": "Kategoria",
        "CityName": null,
        "CountryName": null,
        "RegionIsoCode": null,
        "MetroCode": null,
        "CountryIsoCode": null,
        "RegionName": null
      },
      "Second": {
        "Timestamp": "2016-06-27T00:00:58.599+00:00",
        "IsRobot": true,
        "Channel": "#pl.wikipedia",
        "Flags": "NB",
        "IsUnpatrolled": false,
        "Page": "Kategoria:Dyskusje nad usunięciem artykułu zakończone bez konsensusu − lipiec 2016",
        "DiffUri": "https://pl.wikipedia.org/w/index.php?oldid=46204477&rcid=68522573",
        "Added": 270,
        "Comment": "utworzenie kategorii",
        "CommentLength": 20,
        "IsNew": true,
        "IsMinor": false,
        "Delta": 270,
        "IsAnonymous": false,
        "User": "Beau.bot",
        "DeltaBucket": 200,
        "Deleted": 0,
        "Namespace": "Kategoria",
        "CityName": null,
        "CountryName": null,
        "RegionIsoCode": null,
        "MetroCode": null,
        "CountryIsoCode": null,
        "RegionName": null
      }
    }
  },
  {
    "SegmentId": "wikipedia_2016-06-27T00:00:00.000Z_2016-06-28T00:00:00.000Z_2024-02-17T12:37:41.909Z",
    "Value": {
      "First": {
        "Timestamp": "2016-06-27T00:01:01.364+00:00",
        "IsRobot": true,
        "Channel": "#sh.wikipedia",
        "Flags": "MB",
        "IsUnpatrolled": false,
        "Page": "El Paraíso, Bachíniva",
        "DiffUri": "https://sh.wikipedia.org/w/index.php?diff=29388564&oldid=29282570&rcid=34712921",
        "Added": 0,
        "Comment": "Bot: Automatska zamjena teksta  (-[[Administrativna podjela Meksika|Admin]] +[[Administrativna podjela Meksika|Admi]])",
        "CommentLength": 118,
        "IsNew": false,
        "IsMinor": true,
        "Delta": -1,
        "IsAnonymous": false,
        "User": "Kolega2357",
        "DeltaBucket": -100,
        "Deleted": 1,
        "Namespace": "Main",
        "CityName": null,
        "CountryName": null,
        "RegionIsoCode": null,
        "MetroCode": null,
        "CountryIsoCode": null,
        "RegionName": null
      },
      "Second": {
        "Timestamp": "2016-06-27T00:01:01.364+00:00",
        "IsRobot": true,
        "Channel": "#sh.wikipedia",
        "Flags": "MB",
        "IsUnpatrolled": false,
        "Page": "El Paraíso, Bachíniva",
        "DiffUri": "https://sh.wikipedia.org/w/index.php?diff=29388564&oldid=29282570&rcid=34712921",
        "Added": 0,
        "Comment": "Bot: Automatska zamjena teksta  (-[[Administrativna podjela Meksika|Admin]] +[[Administrativna podjela Meksika|Admi]])",
        "CommentLength": 118,
        "IsNew": false,
        "IsMinor": true,
        "Delta": -1,
        "IsAnonymous": false,
        "User": "Kolega2357",
        "DeltaBucket": -100,
        "Deleted": 1,
        "Namespace": "Main",
        "CityName": null,
        "CountryName": null,
        "RegionIsoCode": null,
        "MetroCode": null,
        "CountryIsoCode": null,
        "RegionName": null
      }
    }
  },
  {
    "SegmentId": "wikipedia_2016-06-27T00:00:00.000Z_2016-06-28T00:00:00.000Z_2024-02-17T12:37:41.909Z",
    "Value": {
      "First": {
        "Timestamp": "2016-06-27T00:01:03.685+00:00",
        "IsRobot": true,
        "Channel": "#sh.wikipedia",
        "Flags": "MB",
        "IsUnpatrolled": false,
        "Page": "El Terco, Bachíniva",
        "DiffUri": "https://sh.wikipedia.org/w/index.php?diff=29388566&oldid=29282572&rcid=34712923",
        "Added": 0,
        "Comment": "Bot: Automatska zamjena teksta  (-[[Administrativna podjela Meksika|Admin]] +[[Administrativna podjela Meksika|Admi]])",
        "CommentLength": 118,
        "IsNew": false,
        "IsMinor": true,
        "Delta": -1,
        "IsAnonymous": false,
        "User": "Kolega2357",
        "DeltaBucket": -100,
        "Deleted": 1,
        "Namespace": "Main",
        "CityName": null,
        "CountryName": null,
        "RegionIsoCode": null,
        "MetroCode": null,
        "CountryIsoCode": null,
        "RegionName": null
      },
      "Second": {
        "Timestamp": "2016-06-27T00:01:03.685+00:00",
        "IsRobot": true,
        "Channel": "#sh.wikipedia",
        "Flags": "MB",
        "IsUnpatrolled": false,
        "Page": "El Terco, Bachíniva",
        "DiffUri": "https://sh.wikipedia.org/w/index.php?diff=29388566&oldid=29282572&rcid=34712923",
        "Added": 0,
        "Comment": "Bot: Automatska zamjena teksta  (-[[Administrativna podjela Meksika|Admin]] +[[Administrativna podjela Meksika|Admi]])",
        "CommentLength": 118,
        "IsNew": false,
        "IsMinor": true,
        "Delta": -1,
        "IsAnonymous": false,
        "User": "Kolega2357",
        "DeltaBucket": -100,
        "Deleted": 1,
        "Namespace": "Main",
        "CityName": null,
        "CountryName": null,
        "RegionIsoCode": null,
        "MetroCode": null,
        "CountryIsoCode": null,
        "RegionName": null
      }
    }
  },
  {
    "SegmentId": "wikipedia_2016-06-27T00:00:00.000Z_2016-06-28T00:00:00.000Z_2024-02-17T12:37:41.909Z",
    "Value": {
      "First": {
        "Timestamp": "2016-06-27T00:01:07.347+00:00",
        "IsRobot": true,
        "Channel": "#ceb.wikipedia",
        "Flags": "NB",
        "IsUnpatrolled": false,
        "Page": "Neqerssuaq",
        "DiffUri": "https://ceb.wikipedia.org/w/index.php?oldid=9563239&rcid=36193146",
        "Added": 4150,
        "Comment": "Paghimo ni bot Greenland",
        "CommentLength": 24,
        "IsNew": true,
        "IsMinor": false,
        "Delta": 4150,
        "IsAnonymous": false,
        "User": "Lsjbot",
        "DeltaBucket": 4100,
        "Deleted": 0,
        "Namespace": "Main",
        "CityName": null,
        "CountryName": null,
        "RegionIsoCode": null,
        "MetroCode": null,
        "CountryIsoCode": null,
        "RegionName": null
      },
      "Second": {
        "Timestamp": "2016-06-27T00:01:07.347+00:00",
        "IsRobot": true,
        "Channel": "#ceb.wikipedia",
        "Flags": "NB",
        "IsUnpatrolled": false,
        "Page": "Neqerssuaq",
        "DiffUri": "https://ceb.wikipedia.org/w/index.php?oldid=9563239&rcid=36193146",
        "Added": 4150,
        "Comment": "Paghimo ni bot Greenland",
        "CommentLength": 24,
        "IsNew": true,
        "IsMinor": false,
        "Delta": 4150,
        "IsAnonymous": false,
        "User": "Lsjbot",
        "DeltaBucket": 4100,
        "Deleted": 0,
        "Namespace": "Main",
        "CityName": null,
        "CountryName": null,
        "RegionIsoCode": null,
        "MetroCode": null,
        "CountryIsoCode": null,
        "RegionName": null
      }
    }
  }
]
